### PR TITLE
Refactor Purchase API - Part 1

### PR DESCRIPTION
**Pull request by negue**
Merge all purchase / buy operations into one `buy`-op, except for the special ones like fortify / orb of rebirth

Todo:
- ~~[ ] only use the only `buy`-op in units tests~~
- [x] move purchase of `quantity` inside each op-method
- [x] replace calls `purchase` and `purchaseWithSpell`  to `buy`

### CODE REVIEW

File: website/common/script/ops/buy/buySpecialSpell.js [hunk 1]
```diff
@@ -9,19 +9,34 @@ import {
...
 
   let item = content.special[key];
   if (!item) throw new NotFound(i18n.t('spellNotFound', {spellId: key}, req.language));
 
-  if (user.stats.gp < item.value) {
+  if (user.stats.gp < item.value * quantity) {
     throw new NotAuthorized(i18n.t('messageNotEnoughGold', req.language));
   }
-  user.stats.gp -= item.value;
+  user.stats.gp -= item.value * quantity;
```
Comment (TheHollidayInn) [inline website/common/script/ops/buy/buySpecialSpell.js]: Are you removing the loop? I'd rather not have the quantity code in multiple places

Comment (negue) [reply website/common/script/ops/buy/buySpecialSpell.js]: yes I removed the loop, I think its much cleaner having the `quantity` inside of each op, instead of having a loop which calls the method per iteration. Its also the better way for performance

Comment (TheHollidayInn) [reply website/common/script/ops/buy/buySpecialSpell.js]: Yea, I can see that. But, this does show a lot of duplication and now is a case of "Shotgun Surgery." So, it would be good to remove that.

Comment (negue) [reply website/common/script/ops/buy/buySpecialSpell.js]: What do you think about a shared method of checking the currency by type, value and quantity and the same for removing the currency ?

Comment (TheHollidayInn) [reply website/common/script/ops/buy/buySpecialSpell.js]: You can check by type, but I think what we should have is an object or class with the value method. Then using one of the various methods of Javascript inheritance, we can return value *quantity on a `getValue` method and subtract that from a user in the buy method. Is the what you are also thinking?

Comment (negue) [reply website/common/script/ops/buy/buySpecialSpell.js]: yes that is planned too :), I just wanted to have the first refactoring to be api save, once thats ok, the inner refactoring starts in a separate PR

Comment (TheHollidayInn) [reply website/common/script/ops/buy/buySpecialSpell.js]: Hmm. Alright :P Well, before we go to the "inner" section, we should get this up to 100% coverage for the api and the common ops.

Comment (negue) [reply website/common/script/ops/buy/buySpecialSpell.js]: Yeah you are right, gonna focus on the coverage first

File: website/common/script/ops/buy/buyMysterySet.js [hunk 2]
```diff
@@ -1,14 +1,18 @@
...
 import each from 'lodash/each';
 import {
   BadRequest,
   NotAuthorized,
   NotFound,
-} from '../libs/errors';
+} from '../../libs/errors';
+
+module.exports = function buyMysterySet (user, req = {}, analytics, _window) {
+  if (typeof window !== 'undefined') {
```
Comment (paglias) [inline website/common/script/ops/buy/buyMysterySet.js]: is this used for tests only? What about definind a dummy window under `global.window` in tests?

File: website/common/script/ops/buy/buyArmoire.js [hunk 3]
```diff
@@ -23,10 +23,6 @@ module.exports = function buyArmoire (user, req = {}, analytics) {
     throw new NotAuthorized(i18n.t('messageNotEnoughGold', req.language));
   }
 
-  if (item.canOwn && !item.canOwn(user)) {
```
Comment (paglias) [inline website/common/script/ops/buy/buyArmoire.js]: why was this check removed? I cannot find it anywhere else

**Review by paglias — CHANGES_REQUESTED** [review]
Left two comments, otherwise looks good!

File: website/common/script/ops/buy/buyMysterySet.js (hunk 2 above)
Comment (negue) [reply website/common/script/ops/buy/buyMysterySet.js]: oh thanks for the hint!, thats much better

File: website/common/script/ops/buy/buyArmoire.js (hunk 3 above)
Comment (negue) [reply website/common/script/ops/buy/buyArmoire.js]: I removed it because `armoire`'s 

```
  canOwn () {
    return true;
  },
```

should I add the check back?

Comment (paglias) [reply website/common/script/ops/buy/buyArmoire.js]: yeah I'd keep it in case we change it in future

File: website/common/script/ops/buy/buyMysterySet.js [hunk 4]
```diff
@@ -27,8 +23,8 @@ module.exports = function buyMysterySet (user, req = {}, analytics, _window) {
     throw new NotFound(i18n.t('mysterySetNotFound', req.language));
   }
 
-  if (typeof _window !== 'undefined' && !req.noConfirm && _window.confirm) { // TODO move to client
-    if (!_window.confirm(i18n.t('hourglassBuyEquipSetConfirm'))) return;
+  if (!req.noConfirm && window.confirm) { // TODO move to client
```
Comment (paglias) [inline website/common/script/ops/buy/buyMysterySet.js]: shouldn't this be window && window.confirm for the node env where window is undefined?

**Review by paglias — CHANGES_REQUESTED** [review]

File: website/common/script/ops/buy/buyMysterySet.js (hunk 4 above)
Comment (negue) [reply website/common/script/ops/buy/buyMysterySet.js]: thats what I tried too, but:
`if (window && window.confirm) ` results in
`2018-01-28T16:07:20.949Z - error: ReferenceError: window is not defined`

Comment (paglias) [reply website/common/script/ops/buy/buyMysterySet.js]: try 

```
if (typeof window !== 'undefined') {
    window.confirm...
}
```

Comment (negue) [reply website/common/script/ops/buy/buyMysterySet.js]: I thought about it a second after I wrote my message haha, but now I think this is obselete, because every `hourglass` purchase has already a `purchase for 1 hourglass` confirmation on the client before even this code is called / executed, I think it can be removed completely

Comment (paglias) [reply website/common/script/ops/buy/buyMysterySet.js]: ah okay!

File: website/common/script/ops/buy/buyMysterySet.js [hunk 5]
```diff
@@ -23,10 +23,6 @@ module.exports = function buyMysterySet (user, req = {}, analytics) {
     throw new NotFound(i18n.t('mysterySetNotFound', req.language));
   }
 
-  if (typeof window !== 'undefined' && !req.noConfirm && window.confirm) { // TODO move to client
```
Comment (TheHollidayInn) [inline website/common/script/ops/buy/buyMysterySet.js:26]: Why is this removed?

File: website/common/script/libs/getItemInfo.js [hunk 6]
```diff
@@ -237,7 +237,7 @@ module.exports = function getItemInfo (user, type, item, officialPinnedItems, la
         notes: item.notes(language),
         value: item.value,
         currency: 'gold',
-        purchaseType: 'potions',
```
Comment (TheHollidayInn) [inline website/common/script/libs/getItemInfo.js:240]: Will this string change affect the client checks at all? I think these hard coded strings are used elsewhere

File: website/client/store/actions/shops.js [hunk 7]
```diff
@@ -82,7 +82,7 @@ export function purchase (store, params) {
 
 export function purchaseMysterySet (store, params) {
   const user = store.state.user.data;
-  let opResult = buyOp(user, {params, noConfirm: true, type: 'mystery'});
```
Comment (TheHollidayInn) [inline website/client/store/actions/shops.js:104]: Why did you remove he noConfirm default?

**Comment by TheHollidayInn** [comment]
This is looking good and working well for me.

File: website/common/script/ops/buy/buyMysterySet.js (hunk 5 above)
Comment (negue) [reply website/common/script/ops/buy/buyMysterySet.js:26]: see discussion at https://github.com/HabitRPG/habitica/pull/9714/files/89f1ad2b07e0115ab494c3529518cb96ebc8d6d3..dd5870903f41d10531c4eca77c6ac4d6815d8055#diff-e92347dc0fb9e7ddffb5912653a501df

File: website/common/script/libs/getItemInfo.js (hunk 6 above)
Comment (negue) [reply website/common/script/libs/getItemInfo.js:240]: it was changed to potion because `purchaseType` is assigned to `type` in 

```
makeGenericPurchase (item, type = 'buyModal', quantity = 1) {
      this.$store.dispatch('shops:genericPurchase', {
        pinType: item.pinType,
        type: item.purchaseType,
        key: item.key,
        currency: item.currency,
        quantity,
      });
```

then `type` is used for the op-method in `buy.js`


TL;DR no purchaseType is never used in a translation or something

**Review by paglias — APPROVED** [review]

**Comment by paglias** [comment]
Marked as tested an approved, but at this point it'll be merged next week

**Comment by SabreCat** [comment]
@paglias @negue I'm not 100% certain how the current conflict in `website/common/script/index.js` should be resolved here. Can one of you take a look?
